### PR TITLE
Add a script to analyze python module dependencies using static analysis

### PR DESCRIPTION
**Pull request by gjoliver**
## Why are these changes needed?

This script can potentially be integrated with CI framework to figure out the exact set of units for python PRs.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by simon-mo** [comment]
quick question, would using [semgrep](https://semgrep.dev/) help reduce the amount of regex and maintenance overhead of this script? or [ast parse](https://docs.python.org/3/library/ast.html#imports) each file and find the `Import` node?

**Comment by gjoliver** [comment]
> quick question, would using [semgrep](https://semgrep.dev/) help reduce the amount of regex and maintenance overhead of this script? or [ast parse](https://docs.python.org/3/library/ast.html#imports) each file and find the `Import` node?

I was trying to not install any external libraries, so ast will be awesome I guess.
Give me like a day to play with it, and we will have much better idea.

**Comment by simon-mo** [comment]
additionally library should be ok here because we just need them in the CI machine, not users machine

**Comment by gjoliver** [comment]
> additionally library should be ok here because we just need them in the CI machine, not users machine

yep. 👌

**Comment by gjoliver** [comment]
ast is perfect. this PR is now much simplified and more reliable.
@simon-mo, please take another look. thanks for the suggestion.

File: ci/travis/py_dep_analysis.py [hunk 1]
```diff
@@ -0,0 +1,381 @@
...
+
+
+def _file_path_to_module_path(f: str):
+    """Return the corresponding module path for a .py file.
+    """
+    dir, fn = os.path.split(f)
+    return _full_module_path(_bazel_path_to_module_path(dir), fn)
+
+
+def _depends(graph: DepGraph, visited: Dict[int, bool], tid: int, qid: int):
```
Comment (simon-mo) [inline ci/travis/py_dep_analysis.py:235]: need return type annotation

File: ci/travis/py_dep_analysis.py [hunk 2]
```diff
@@ -0,0 +1,381 @@
...
+        visited[c] = True
+        # Reduce to a question of whether there is a path from c to qid.
+        ds = _depends(graph, visited, c, qid)
+        if ds:
+            # From tid -> c -> qid.
+            return [tid] + ds
+    return []
+
+
+def test_depends_on_file(graph: DepGraph, test: str, path: str):
```
Comment (simon-mo) [inline ci/travis/py_dep_analysis.py:257]: return type annotation and doc

**Review by simon-mo — APPROVED** [review]
Nice! Let's run them in CI. Can you add a BUILD file and run this in one of the jobs in .buildkite/pipeline.yml? Maybe even try to run it in the CI for RLlib? (just print out the dep graph but not using it)

File: ci/travis/py_dep_analysis.py (hunk 1 above)
Comment (gjoliver) [reply ci/travis/py_dep_analysis.py]: done

File: ci/travis/py_dep_analysis.py (hunk 2 above)
Comment (gjoliver) [reply ci/travis/py_dep_analysis.py]: done

**Comment by gjoliver** [comment]
@simon-mo, I added return annotations for all the functions.
and added a dry run of py_dep_analysis in determine_tests_to_run.

however, can you share how and where does determine_tests_to_run run on our CI machine?
py_dep_analysis currently assumes we are running it from ray root dir. so if that's not the case, I need to set the proper base_dir etc.

thanks.

**Comment by simon-mo** [comment]
@mwtian FYI. This might be another way of enhancing bazel to do conditional tests.

**Pull request merged**